### PR TITLE
Java ice adapter

### PR DESCRIPTION
**Pull request by Geosearchef**

### CODE REVIEW

File: src/main/java/com/faforever/client/fa/relay/GpgClientCommand.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
   REHOST("Rehost"),
   DESYNC("Desync"),
   GAME_FULL("GameFull"),
-  ENDED("Ended"),
+  GAME_ENDED("GameEnded"),//TODO: it's called GameEnded
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/GpgClientCommand.java]: What needs to be done here, exactly? And comments should be put on the line above, not behind

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 2]
```diff
@@ -57,7 +56,8 @@
 @Slf4j
 public class IceAdapterImpl implements IceAdapter, InitializingBean, DisposableBean {
 
-  private static final int CONNECTION_ATTEMPTS = 5;
+  private static final int CONNECTION_ATTEMPTS = 50;
+  private static final int CONNECTION_ATTEMPT_DELAY = 100;//MS
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: Rather call the variable `CONNECTION_ATTEMPT_DELAY_MILLIS`

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterApi.java [hunk 3]
```diff
@@ -42,5 +42,5 @@
    * ICE server array for use in webrtc. Must be called before joinGame/connectToPeer. See <a
    * href="https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer">https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer</a>.
    */
-  String setIceServers(List<Map<String, String>> iceServers);
+  void setIceServers(List<Map<String, Object>> iceServers);
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterApi.java:47]: Consider documenting the parameter.

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 4]
```diff
@@ -93,25 +93,28 @@ public void afterPropertiesSet() {
   }
 
   @SneakyThrows
-  private List<Map<String, String>> toIceServers(List<IceServersServerMessage.IceServer> iceServers) {
-    return iceServers.stream()
-        .map(this::toIceServer)
-        .collect(Collectors.toList());
-  }
+  private List<Map<String, Object>> toIceServers(List<IceServersServerMessage.IceServer> iceServers) {
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java:109]: Consider documenting the return values.

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 5]
```diff
@@ -155,11 +158,15 @@ public void onGpgGameMessage(GpgGameMessageEvent event) {
...
+          Paths.get(System.getProperty("java.home")).resolve("bin").resolve(org.bridj.Platform.isWindows() ? "java.exe" : "java").toAbsolutePath().toString(),
+          "-jar",
           getBinaryName(workDirectory),
           "--id", String.valueOf(currentPlayer.getId()),
           "--login", currentPlayer.getUsername(),
           "--rpc-port", String.valueOf(adapterPort),
-          "--gpgnet-port", String.valueOf(gpgPort)
+          "--gpgnet-port", String.valueOf(gpgPort),
+          "--log-directory", "iceAdapterLogs",
+          "--debug-window"
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: `--debug-window` probably isn't desired for production, should be configurable, e.g. via environment variable

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 6]
```diff
@@ -184,7 +191,14 @@ public void onGpgGameMessage(GpgGameMessageEvent event) {
             setLobbyInitMode();
             break;
           } catch (ConnectException e) {
-            logger.debug("Could not connect to ICE adapter (attempt {}/{})", attempt, CONNECTION_ATTEMPTS);
+            logger.debug("Could not connect to ICE adapter (attempt {}/{})", attempt + 1, CONNECTION_ATTEMPTS);
+          }
+
+          //Wait as the socket fails too fast on unix/linux not giving the adapter enough time to start
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: Space after `//` :P

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 7]
```diff
@@ -184,7 +191,14 @@ public void onGpgGameMessage(GpgGameMessageEvent event) {
...
+          try {
+            Thread.sleep(CONNECTION_ATTEMPT_DELAY);
+          } catch (InterruptedException e) {
+            logger.error("Error while waiting for ice adapter", e);
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: `warn`

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 8]
```diff
@@ -207,7 +221,8 @@ public void onGpgGameMessage(GpgGameMessageEvent event) {
   }
 
   private String getBinaryName(Path workDirectory) {
-    return Platform.isWindows() ? workDirectory.resolve("faf-ice-adapter.exe").toString() : "./faf-ice-adapter";
+//    return Platform.isWindows() ? workDirectory.resolve("faf-ice-adapter.exe").toString() : "./faf-ice-adapter";
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: Remove dead code

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java [hunk 9]
```diff
@@ -242,7 +257,7 @@ private IceAdapterApi newIceAdapterProxy() {
             log.warn("Ignoring call to ICE adapter as we are not connected: {}({})", method.getName(), argList);
             return null;
           }
-          log.debug("Calling {}({})", method.getName(), argList);
+          log.warn("Calling {}({})", method.getName(), argList);
```
Comment (micheljung) [inline src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: That's nothing to `warn` about

**Review by micheljung — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 5 above)
Comment (Geosearchef) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: for now that's desired. We will probable have to release another update once ICE is running stable. I will change the debug window though (e.g. tray icon) so that we don't scare the users.

The reason for keeping this is having fast access at it when stuff goes wrong over the next days and especially as I've sometimes seen the adapter not terminating so people can close it.

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 9 above)
Comment (Geosearchef) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: I think this should be a warning, this is not usual behaviour?

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 5 above)
Comment (Geosearchef) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: This will stay for now (till the adapter is stable/proofed to be working), but the ice adapter will change so that the debug window will be a tray icon or small info window by default.

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 9 above)
Comment (micheljung) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: Calling a method on the ICE adapter is usual behavior

File: downlords-faf-client.install4j [hunk 10]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<install4j version="7.0.7" transformSequenceNumber="7">
-  <directoryPresets config="./src/media/appicon" />
+<install4j version="7.0.6" transformSequenceNumber="7">
+  <directoryPresets config="./build/resources" />
```
Comment (micheljung) [inline downlords-faf-client.install4j]: Was this change intended?

File: downlords-faf-client.install4j [hunk 11]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<install4j version="7.0.7" transformSequenceNumber="7">
-  <directoryPresets config="./src/media/appicon" />
+<install4j version="7.0.6" transformSequenceNumber="7">
```
Comment (micheljung) [inline downlords-faf-client.install4j]: We should rather go forward than backward ;-)

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 9 above)
Comment (Geosearchef) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: oh fu, euhm, yes, that is, sry, I mistook this for being part of the not connected case above.

File: gradle.properties [hunk 12]
```diff
@@ -19,3 +19,4 @@ javafxVersion=11
 coverallsGradlePluginVersion=2.8.2
 jnaVersion=5.0.0
 jacocoVersion=0.8.2
+install4jHomeDir=/home/geosearchef/Programs/install4j7
```
Comment (micheljung) [inline gradle.properties]: Does not belong here

**Review by micheljung — DISMISSED** [review]

File: downlords-faf-client.install4j (hunk 11 above)
Comment (Geosearchef) [reply downlords-faf-client.install4j]: I think the entire install4j file can be reverted

File: downlords-faf-client.install4j (hunk 10 above)
Comment (Geosearchef) [reply downlords-faf-client.install4j]: no, install4j did that automatically

**Review by 1-alex98 — DISMISSED** [review]

File: src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java (hunk 5 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/client/fa/relay/ice/IceAdapterImpl.java]: Here is a solution that is enabled by default but configurable: https://github.com/FAForever/downlords-faf-client/commit/ca51b1dc3a845cd6b371b6b04994a2b9287dc719

**Comment by 1-alex98** [comment]
@Geosearchef what is this i thought you merged the ice adapter changes pls close if outdated

**Comment by Geosearchef** [comment]
No, it wasn't merged, all releases are built off this branch currently

**Review by 1-alex98 — APPROVED** [review]

**Comment by 1-alex98** [comment]
@Geosearchef then why dont we merge? Do you want to port it to v2 branch? Or are you missing the rights to?

**Comment by Geosearchef** [comment]
Downlord didn't want to merge it (yet?)

**Comment by 1-alex98** [comment]
Well then i will merge it. It further blocks everything. I can not work like this.... If we have yet another develop branch
